source/ui/netsender.c

```diff
@@ -0,0 +1,537 @@
+	z_stream strm;
+
+	/* allocate deflate state */
+	strm.zalloc = Z_NULL;
+	strm.zfree = Z_NULL;
+	strm.opaque = Z_NULL;
+	ret = deflateInit(&strm, Z_DEFAULT_COMPRESSION);
+	if (ret != Z_OK) return;
+
+	datafd = socket(AF_INET,SOCK_STREAM,0);
+	if (datafd < 0)
+	{
+		netsenderError("datafd", errno);
+		return;
+	}
+
+	struct sockaddr_in s;
+	memset(&s, 0, sizeof(struct sockaddr_in));
+	s.sin_family = AF_INET;
+	s.sin_port = htons(NETWORK_PORT);
+	s.sin_addr.s_addr = inaddr;
+
+	if (connect(datafd, (struct sockaddr *)&s, sizeof(s)) < 0 )
+	{
+		netsenderError("connect", errno);
+		goto send3DSXFileExit;
+	}
+
+	int namelen = strlen(name);
+
+	if (sendInt32LE(datafd, namelen))
+	{
+		netsenderError("filenameLen", errno);
+		goto send3DSXFileExit;
+	}
+
+	if (sendData(datafd, namelen, name))
+	{
+		netsenderError("filename", errno);
+		goto send3DSXFileExit;
+	}
+
+	if (sendInt32LE(datafd, filelen))
+	{
+		netsenderError("filelen", errno);
+		goto send3DSXFileExit;
+	}
+
+	s32 response;
+	if (recvInt32LE(datafd, &response) != 0)
+	{
+		netsenderError("response", 0);
+		goto send3DSXFileExit;
+	}
+
+	if (response != 0)
+	{
+		switch (response) {
+			case -1:
+				netsenderError("create file", 0);
+				break;
+			case -2:
+				netsenderError("no space", 0);
+				break;
+			case -3:
+				netsenderError("no memory", 0);
+				break;
+			default:
+				netsenderError("response:", response);
+				break;
+		}
+		goto send3DSXFileExit;
+	}
+
+	size_t totalsent = 0, blocks = 0;
+
+	do
+	{
+		strm.avail_in = fread(in, 1, ZLIB_CHUNK, fh);
+		filetotal += strm.avail_in;
+		if (ferror(fh))
+		{
+			netsenderError("ferror", 0);
+			deflateEnd(&strm);
+			goto send3DSXFileExit;
+		}
+		flush = feof(fh) ? Z_FINISH : Z_NO_FLUSH;
+		strm.next_in = in;
+		/* run deflate() on input until output buffer not full, finish
+		   compression if all of source has been read in */
+		do
+		{
+			strm.avail_out = ZLIB_CHUNK;
+			strm.next_out = out;
+			ret = deflate(&strm, flush);    /* no bad return value */
+			have = ZLIB_CHUNK - strm.avail_out;
+
+			if (have != 0)
+			{
+				if (sendInt32LE(datafd,have))
+				{
+					netsenderError("chunk size", errno);
+					goto send3DSXFileExit;
+				}
+
+				if (sendData(datafd, have, (char*)out))
+				{
+					netsenderError("have", errno);
+					deflateEnd(&strm);
+					goto send3DSXFileExit;
+				}
+
+				totalsent += have;
+				blocks++;
+			}
+		}
+		while (strm.avail_out == 0);
+		/* done when last data in file processed */
+	}
+	while (flush != Z_FINISH);
+	deflateEnd(&strm);
+
+	if (recvInt32LE(datafd,&response)!=0)
+		goto send3DSXFileExit;
+
+	static char cmdbuf[3072];
+	memset(cmdbuf, 0, 3072);
+	snprintf(&cmdbuf[4], 3072-4, "3dslink:/%s", name);
+	u32 cmdlen = strlen(&cmdbuf[4])+1;
+
+	cmdbuf[0] = cmdlen & 0xff;
+	cmdbuf[1] = (cmdlen>>8) & 0xff;
+	cmdbuf[2] = (cmdlen>>16) & 0xff;
+	cmdbuf[3] = (cmdlen>>24) & 0xff;
+
+	if (sendData(datafd,cmdlen+4,cmdbuf))
+		netsenderError("cmdbuf", errno);
+	
+	send3DSXFileExit:
```

Comment:
Labels aren't indented, should be on column 0. Also, I'd call this label `_cleanup:`